src/Workspaces/CSharp/Portable/Extensions/SemanticModelExtensions.cs

```diff
@@ -451,9 +451,47 @@ public static ISet<INamespaceSymbol> GetUsingNamespacesInScope(this SemanticMode
                 }
             }
 
+            // For type constraints use declared accessibility of its member declaration (class, interface, delegate, method)
+            if (type.IsParentKind(SyntaxKind.TypeConstraint) &&
+                type.Parent.IsParentKind(SyntaxKind.TypeParameterConstraintClause))
+            {
+                var memberDeclaration = type.GetAncestor<MemberDeclarationSyntax>();
+                var declaredAccessibility = semanticModel.GetDeclaredSymbol(memberDeclaration, cancellationToken).DeclaredAccessibility;
+
+                return declaredAccessibility;
+            }
+
             return Accessibility.Private;
         }
 
+        public static bool AllContainingTypesArePublic(
+            this SemanticModel semanticModel,
+            TypeSyntax type,
+            CancellationToken cancellationToken)
+        {
+            if (type == null)
+            {
+                return false;
+            }
+
+            type = GetOutermostType(type);
```

Comment:
Same type is passed to `DetermineAccessibilityConstraint` which has `type = GetOutermostType(type);`